{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Introduction\n",
    "\n",
    "This is a short tutorial for conducting a conditional analysis using PLINK2 and Ricopili. It is based on the analysis done for the PGC 2017 Bipolar GWAS project.\n",
    "\n",
    "The full `snakemake` pipeline can be found in the `pipelines/` directory at the root of this repository. If you are familiar with `make` or `snakemake`, it should be easy to adapt this directly to your analysis. \n",
    "\n",
    "Otherwise, below is a breakdown of the various steps of the analysis. File locations are replaced with variables in brackets, i.e. `/path/to/input/plink.bed` will be shown as `{input.bed}` in the code. If a step shows multiple inputs, the input object is assumed to be a list of strings. Outputs from one step of the workflow are defined inputs to the next (for dependencies between steps, see workflow diagram below).\n",
    "\n",
    "As a high level view, we will be re-creating the following example conditional analysis given two studies and two loci of interest:\n",
    "![workflow_diagram](files/resources/conditional_analysis_dag.png \"workflow_diagram\")\n",
    "\n",
    "Each of the steps in this workflow are described as sections below.\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Requirements\n",
    "The following dependencies are assumed to be installed:\n",
    "  * PLINK2\n",
    "  * Ricopili area plot script\n",
    "  * Python3 with packages:\n",
    "      - Pandas\n",
    "      - snakemake"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Starting data\n",
    "## Genetic Data\n",
    "We assume that you have a set of studies in PLINK binary format that you wish to conduct a conditional analysis in.\n",
    "\n",
    "## Defined Loci\n",
    "Additionally, we assume that there exists a tab-separated file (.tsv) that defines loci of interest with the following labeled columns:\n",
    " * SNP : The variant ID of the index SNP you with to condition on.\n",
    " * CHR : The chromosome of the same index SNP.\n",
    " * left : The start (in base pairs) of the locus containing the index SNP. Used to subset the input data and in generating the area plot.\n",
    " * right : The end (in base pairs) of the locus containing the index SNP.\n",
    "\n",
    "## Covariate File\n",
    "We use the `mds_cov` file generated by the Ricopili PCA pipeline.\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Workflow"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## make_regions_file: Define loci of interest in a file\n",
    "We will define the loci in a file compatible with PLINK's extract functionality.\n",
    "\n",
    "```\n",
    "import pandas as pd\n",
    "\n",
    "locus_file = os.path.join(analysis_dir, 'data/Jun2017_withinLocusCond_coordinate_vasa_corrections.tsv')\n",
    "locus_dat = pd.read_csv(locus_file, sep='\\t', header=0)\n",
    "\n",
    "with open(output.regions_file, 'w') as set_file:\n",
    "    for index, locus in locus_dat.iterrows():\n",
    "        line = '\\t'.join([str(x) for x in [locus.CHR, locus.left, locus.right, locus.SNP]])\n",
    "        set_file.write(line + '\\n')\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## extract_variants: Subset input data to loci of interest\n",
    "Subset your input data to variants in the loci of interest.\n",
    "\n",
    "For each study, you will have a file path root to the input PLINK binary files. For each study, you will fill out and execute the following command:\n",
    "```\n",
    "from snakemake import shell\n",
    "\n",
    "shell('plink '\n",
    "        '--bfile {input_prefix} '\n",
    "        '--extract range {input.regions_file} '\n",
    "        '--make-bed '\n",
    "        '--out {output_prefix}')\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## merge: Merge individual studies into a single dataset\n",
    "Write out a the list of PLINK files that have the loci extracted. Then use PLINK to merge these datasets:\n",
    "```\n",
    "with open({output.merge_list}, 'w') as merge_list:\n",
    "    for bed in {input.beds}:\n",
    "        if os.path.getsize(bed) > 1:\n",
    "            # ignore studies with no SNPs in ranges\n",
    "            prefix = bed.replace('.bed', '')\n",
    "            merge_list.write(prefix + '\\n')\n",
    "\n",
    "output_prefix = output.bed.replace('.bed', '')\n",
    "shell('plink '\n",
    "        '--merge-list {output.merge_list} '\n",
    "        '--out {output_prefix}')\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## make_covariate_file: Create the Covariate file\n",
    "We are going to remove uncesessary covariates from the `mds_cov` file. We end up with covariates for principal components (significantly associated with the outcome) and study indicators (for non-emtpy studies):\n",
    "```\n",
    "mds_dat = pd.read_csv({mds_cov}, sep='\\s+')\n",
    "\n",
    "# remove study columns with all zeros -- these are trio studies\n",
    "study_indicator = [x for x in mds_dat.columns if x.startswith('st')]\n",
    "mds_dat = mds_dat.loc[:, (mds_dat != 0).any(axis=0)]\n",
    "\n",
    "# subset to signifiant PCs for easy access later\n",
    "pcs = set([x for x in mds_dat.columns if x.startswith('C')])\n",
    "significant_pcs = set(['C1', 'C2', 'C3', 'C4', 'C5', 'C6', 'C19'])\n",
    "non_significant_pcs = pcs.difference(significant_pcs)\n",
    "mds_dat.drop(non_significant_pcs, axis=1, inplace=True)\n",
    "\n",
    "mds_dat.to_csv(output.covariate_file, sep='\\t', index=False)\n",
    "```\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## keep_covar_samples: Subset to samples in the modified covariate file\n",
    "..."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## make_condition_list: Create list of variants to condition on\n",
    "..."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## conditional_analysis: Perform conditional analysis with PLINK\n",
    "..."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## filter_conditional_assoc_results: Filter association results\n",
    "..."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## conditional_region_plot_setup: Setup region plotting script\n",
    "..."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## conditional_region_plots: Generate conditional region plots\n",
    "..."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.5.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 0
}
